server/src/main/java/org/elasticsearch/index/engine/TranslogDirectoryReader.java

```diff
@@ -159,20 +191,55 @@ static DirectoryReader createInMemoryReader(
             IndexWriterConfig.OpenMode.CREATE
         ).setCodec(engineConfig.getCodec());
         try (IndexWriter writer = new IndexWriter(directory, writeConfig)) {
-            writer.addDocument(parsedDocs.rootDoc());
+            final int numDocs;
+            if (rootDocOnly) {
+                numDocs = 1;
+                writer.addDocument(parsedDocs.rootDoc());
+            } else {
+                numDocs = parsedDocs.docs().size();
+                writer.addDocuments(parsedDocs.docs());
```

Comment:
I am not 100% sure I understand why this is necessary for synthetic source and not for regular source, perhaps you can help me out?

Reply:
Synthesizing the source of nested fields requires both the root document and its child documents, whereas the stored source only requires the root document: https://github.com/elastic/elasticsearch/blob/12e86b1cd05202cdc080ed1ad74438c041d531de/server/src/main/java/org/elasticsearch/index/mapper/NestedObjectMapper.java#L466

Reply:
Thanks. I wonder about the field fetch support that is in the GET API, but I suppose it is restricted to not be able to access nested objects in any way?

Reply:
> I wonder about the field fetch support that is in the GET API, but I suppose it is restricted to not be able to access nested objects in any way?

I don't think including stored fields of nested (child) documents works today. Looks like the root docid is used in `ShardGetService#innerGetFetch(...)`.